Revisit Discovery Provider Overview Dashboards with additional celery metrics

joaquincasares opened the pull request:
### Description 

Fix existing Grafana panels. This PR:

* Resets metrics for completed/non-active celery tasks by using the difference between the active and registered Celery tasks.
* Saves duration metrics for completed (successful or failed) celery tasks using a decorator that's now applied to all celery_tasks.
* New dashboards to expose the new metrics.
* Various smaller fixes for our Grafana dev workflow.

### Tests

Locally check metric values:

```
$ curl -s "http://localhost:5000/prometheus_metrics"
```

Build charts with confirmed metrics using `A run monitoring up`.

### Monitoring - How will this change be monitored? Are there sufficient logs / alerts?

![screencapture-34-132-90-76-d-Ayvzmeq7z-audius-discovery-provider-overview-2022-07-06-17_51_39](https://user-images.githubusercontent.com/545666/177656356-3eb6ed5e-e945-4134-b1ef-907fbcc3cd3f.png)

![screencapture-34-132-90-76-d-ZUjg533nk-audius-discovery-provider-celery-tasks-2022-07-06-17_52_25](https://user-images.githubusercontent.com/545666/177656355-65fcafda-3a50-4455-b0ab-77505a111c45.png)

![screencapture-34-132-90-76-d-W9FaXK6nk-audius-discovery-provider-celery-task-details-2022-07-06-17_52_49](https://user-images.githubusercontent.com/545666/177656352-9da83660-d4cb-496b-b8f0-35065a9203f5.png)

dmanjunath commented on discovery-provider/src/utils/prometheus_metric.py:
```diff
@@ -7,6 +8,28 @@
+            )
+            try:
+                result = func(*args, **kwargs)
+                metric.save_time({"func_name": func.__name__, "success": True})
```
func.__name__ is the name of the @celery.task?

joaquincasares replied:
```
@celery.task(name="index_trending", bind=True)
@save_duration_metric(metric_group="celery_task")
def index_trending_task(self):
```

In the above example, `func.__name__` will be `index_trending_task`.

This is because:

* decorators are applied from the bottom up,
* the `@celery.task` must be the last decorator applied,
* `@save_duration_metric` will only have access to the caller function.

dmanjunath commented on discovery-provider/src/utils/prometheus_metric.py:
```diff
@@ -7,6 +8,28 @@
 logger = logging.getLogger(__name__)
 
 
+def save_duration_metric(metric_group):
```
might be nice to add some comments here

dmanjunath commented on discovery-provider/src/monitors/monitoring_queue.py:
```diff
@@ -41,6 +42,7 @@ def refresh(redis, db, monitor):
 
 
 @celery.task(name="monitoring_queue", bind=True)
+@save_duration_metric(metric_group="celery_task")
```
how confident are you about this not introducing regressions?

joaquincasares replied:
Fairly confident since we're merely holding onto the `result`, saving a metric, then returning the `result`.

Here is the bulk of the actionable code:

```
            try:
                result = func(*args, **kwargs)
                metric.save_time({"func_name": func.__name__, "success": True})
                return result
            except:
                metric.save_time({"func_name": func.__name__, "success": False})
                raise
```

dmanjunath commented:
one more thing is it would be nice to have a registry of metric keys and some comments about what they are like in content node. it's hard to track that in code if they're done ad hoc everywhere

joaquincasares commented on discovery-provider/src/monitors/celery.py:
```diff
@@ -47,4 +48,23 @@ def get_celery_tasks(**kwargs) -> str:
                 continue
 
-    return json.dumps(celery_tasks)
+    registered_celery_tasks = []
```
All tasks that are not currently running:

* https://docs.celeryq.dev/en/stable/reference/celery.app.control.html#celery.app.control.Inspect.registered_tasks

joaquincasares commented:
> one more thing is it would be nice to have a registry of metric keys and some comments about what they are like in content node. it's hard to track that in code if they're done ad hoc everywhere

Sure, I've been thinking about that, but let's complete that work under INF-164.

dmanjunath commented on discovery-provider/src/monitors/celery.py:
```diff
@@ -28,6 +28,7 @@ def get_celery_tasks(**kwargs) -> str:
     i: Inspect = celery_app.control.inspect()
 
     active = i.active()
```
@joaquincasares how is it not acting appropriately?

joaquincasares replied:
![Screen Shot 2022-07-06 at 4 26 07 PM](https://user-images.githubusercontent.com/545666/177646332-1054bf43-19ed-4534-8b7e-12b330316d9a.png)

The Active chart keeps crawling from 0 to 120s, then resets back to 0, even though we don't record any successful nor failed tasks that take 120s.

This leads me to believe that Active just lets us know the "active queues" that are being processed, not the "active tasks" that are processing.

dmanjunath commented on discovery-provider/src/monitors/celery.py:
```diff
@@ -47,4 +48,23 @@ def get_celery_tasks(**kwargs) -> str:
+                continue
+
+    ret = {
+        "celery_tasks": celery_tasks,
```
might be better called active_tasks

dmanjunath commented on monitoring/grafana/bin/save-dashboards.sh:
```diff
@@ -27,7 +27,8 @@ CLEAR_DASHBOARD_ID='.id = null'
 # CLEAR PROMETHEUS UID
 # clears prometheus uid since each deployment is unique
-CLEAR_PROM_UID='del(.panels[].targets[]?.datasource.uid)'
+CLEAR_PROM_TARGET_UID='del(.panels[].targets[]?.datasource.uid)'
```
not part of the scope of this feature but have you seen this? https://github.com/mpostument/grafana-sync

joaquincasares replied:
No, I haven't, thanks!

I added it as a comment to INF-169 since it might help with our folder support.

dmanjunath commented on discovery-provider/src/utils/prometheus_metric.py:
```diff
@@ -7,6 +8,63 @@
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            histogram_metric = PrometheusMetric(
```
why have both?

joaquincasares replied:
@dmanjunath Mentioned on lines 13 and 14, but mainly you can't gain raw seconds from histograms.

Here's a bit of info on [errors of quantile estimation](https://prometheus.io/docs/practices/histograms/#errors-of-quantile-estimation).

dmanjunath reviewed (approved):
overall this looks good, just had a few small questions

joaquincasares commented:
> also @joaquincasares mind cleaning up the PR title and description a bit more and adding more detail? for those that didn't run the local stack and hit the endpoint it would be nice to have JSON responses and pictures of the expected dashboard layouts

Sure thing! I've cleaned up the PR title and description.

I've also added screenshots for the new dashboards.

All JSON versions of the dashboard can always be found within `grafana/dashboards`.
